File: benchmarks/lua/compute/fannkuch.lua
-- The Computer Language Benchmarks Game
-- https://salsa.debian.org/benchmarksgame-team/benchmarksgame/
-- Fannkuch-redux benchmark

local N = tonumber(arg and arg[1]) or 10

local perm = {}
local perm1 = {}
local count = {}
local maxflips = 0
local checksum = 0

-- Initialize
for i = 0, N - 1 do
    perm1[i] = i
end

local r = N

local function flip_count()
    -- Copy perm1 into perm
    for i = 0, N - 1 do
        perm[i] = perm1[i]
    end

    local flips = 0
    local first = perm[0]
    while first ~= 0 do
        -- Reverse perm[0..first]
        local lo, hi = 0, first
        while lo < hi do
            perm[lo], perm[hi] = perm[hi], perm[lo]
            lo = lo + 1
            hi = hi - 1
        end
        flips = flips + 1
        first = perm[0]
    end
    return flips
end

local perm_count = 0

while true do
    -- Use current permutation
    while r ~= 1 do
        count[r - 1] = r
        r = r - 1
    end

    local flips = flip_count()
    if flips > maxflips then
        maxflips = flips
    end
    if perm_count % 2 == 0 then
        checksum = checksum + flips
    else
        checksum = checksum - flips
    end
    perm_count = perm_count + 1

    -- Generate next permutation
    local done = false
    while true do
        if r == N then
            done = true
            break
        end

        -- Rotate perm1[0..r] left by one
        local perm0 = perm1[0]
        for i = 0, r - 1 do
            perm1[i] = perm1[i + 1]
        end
        perm1[r] = perm0

        count[r] = count[r] - 1
        if count[r] > 0 then
            break
        end
        r = r + 1
    end

    if done then
        break
    end
end

io.write(string.format("%d\nPfannkuchen(%d) = %d\n", checksum, N, maxflips))

File: benchmarks/js/compute/fannkuch.js
// Fannkuch-redux benchmark
// Print checksum and max flips count

var print = typeof console !== 'undefined' ? console.log.bind(console) : globalThis.print;
var DEFAULT = 10;
var N = parseInt(typeof scriptArgs !== 'undefined' ? scriptArgs[1] : (typeof process !== 'undefined' ? process.argv[2] : '0')) || DEFAULT;

function fannkuch(n) {
    var perm = new Array(n);
    var perm1 = new Array(n);
    var count = new Array(n);
    var maxFlipsCount = 0;
    var checksum = 0;
    var permCount = 0;

    for (var i = 0; i < n; i++) {
        perm1[i] = i;
    }

    var r = n;
    while (true) {
        while (r !== 1) {
            count[r - 1] = r;
            r--;
        }

        for (var i = 0; i < n; i++) {
            perm[i] = perm1[i];
        }

        var flipsCount = 0;
        var k;

        while ((k = perm[0]) !== 0) {
            var k2 = (k + 1) >> 1;
            for (var i = 0; i < k2; i++) {
                var temp = perm[i];
                perm[i] = perm[k - i];
                perm[k - i] = temp;
            }
            flipsCount++;
        }

        if (flipsCount > maxFlipsCount) {
            maxFlipsCount = flipsCount;
        }

        if (permCount % 2 === 0) {
            checksum += flipsCount;
        } else {
            checksum -= flipsCount;
        }
        permCount++;

        // Generate next permutation
        while (true) {
            if (r === n) {
                return [checksum, maxFlipsCount];
            }

            var perm0 = perm1[0];
            var i = 0;
            while (i < r) {
                var j = i + 1;
                perm1[i] = perm1[j];
                i = j;
            }
            perm1[r] = perm0;

            count[r]--;
            if (count[r] > 0) {
                break;
            }
            r++;
        }
    }
}

var result = fannkuch(N);
print(result[0]);
print("Pfannkuchen(" + N + ") = " + result[1]);
